### PR TITLE
Enabling DeepSMILE on large encoded datasets

### PR DESCRIPTION
**Pull request by vale-salvatelli**
This PR contains two changes necessary to run DeepSMILE on a large dataset when using the Innereye SSL checkpoint (or any other encoder with high dimension):
* option to encode in chunks (this prevents OOM error when performing the encoding)
* option to load the cached encoded dataset in CPU (this prevents OOM when loading from the cache)

It also changes how the PNG images are loaded all over the histo pipeline to make the loading faster (see https://hi-ml.readthedocs.io/en/latest/loading_images.html)

Please follow the guidelines for PRs contained [here](docs/pull_requests.md). Checklist:

- [ x] Ensure that your PR is small, and implements one change.
- [ x] Add unit tests for all functions that you introduced or modified.
- [ x] Run PyCharm's code cleanup tools on your Python files.
- [ ] Link the correct GitHub issue for tracking.
- [x ] Update the [Changelog](CHANGELOG.md) file: Describe your change in terms of 
Added/Changed/Removed/... in the "Upcoming" section.
- [ ] When merging your PR, replace the default merge message with a description of your PR,
and if needed a motivation why that change was required.

### CODE REVIEW

File: InnerEye/ML/Histopathology/datamodules/base_module.py [hunk 1]
```diff
@@ -46,16 +50,25 @@ def __init__(self, root_path: Path, max_bag_size: int = 0, batch_size: int = 1,
...
+          - `NONE (default)`: no pre-cache is performed;
         :param cache_dir: The directory onto which to cache data if caching is enabled.
         :param number_of_cross_validation_splits: Number of folds to perform.
         :param cross_validation_split_index: Index of the cross validation split to be performed.
         """
+        if precache_location == CacheLocation.NONE:
+            save_precache = None
+        else:
+            save_precache = precache_location
+
```
Comment (ant0nsc) [inline InnerEye/ML/Histopathology/datamodules/base_module.py:71]: If this is how you are using the CacheLocation enum, I would find it cleaner to have the arguments defined as Optional[CacheLocation], and remove the NONE enum value.

File: InnerEye/ML/Histopathology/datamodules/base_module.py [hunk 2]
```diff
@@ -87,17 +100,20 @@ def prepare_data(self) -> None:
             self._load_dataset(self.val_dataset, stage='val', shuffle=True)
             self._load_dataset(self.test_dataset, stage='test', shuffle=True)
 
-    def _dataset_pickle_path(self, stage: str) -> Optional[Path]:
+    def _dataset_pt_path(self, stage: str) -> Optional[Path]:
```
Comment (ant0nsc) [inline InnerEye/ML/Histopathology/datamodules/base_module.py:103]: what's PT? maybe spell out?

File: InnerEye/ML/Histopathology/datamodules/base_module.py [hunk 3]
```diff
@@ -87,17 +100,20 @@ def prepare_data(self) -> None:
...
+        return self.cache_dir / f"{stage}_dataset.pt"
 
     def _load_dataset(self, tiles_dataset: TilesDataset, stage: str, shuffle: bool) -> Dataset:
-        dataset_pickle_path = self._dataset_pickle_path(stage)
+        dataset_pt_path = self._dataset_pt_path(stage)
 
-        if dataset_pickle_path and dataset_pickle_path.exists():
-            with dataset_pickle_path.open('rb') as f:
-                return pickle.load(f)
+        if dataset_pt_path and dataset_pt_path.exists():
```
Comment (ant0nsc) [inline InnerEye/ML/Histopathology/datamodules/base_module.py:111]: .exists() -> .is_file()

File: InnerEye/ML/configs/histo_configs/classification/BaseMIL.py [hunk 4]
```diff
@@ -44,8 +44,12 @@ class BaseMIL(LightningContainer):
     cache_mode: CacheMode = param.ClassSelector(default=CacheMode.MEMORY, class_=CacheMode,
                                                 doc="The type of caching to perform: "
                                                     "'memory' (default), 'disk', or 'none'.")
-    save_precache: bool = param.Boolean(True, doc="Whether to pre-cache the entire transformed "
-                                                  "dataset upfront and save it to disk.")
+    precache_location: str = param.ClassSelector(default=CacheLocation.NONE, class_=CacheLocation,
+                                                 doc="Whether to pre-cache the entire transformed dataset upfront "
+                                                 "and save it to disk and if re-load in cpu or gpu. Options:"
+                                                 "`none` (default),`cpu`, `gpu`")
```
Comment (ant0nsc) [inline InnerEye/ML/configs/histo_configs/classification/BaseMIL.py:50]: see my comment above, maybe make that optional, rather than encoding None in the enum?

File: InnerEye/ML/configs/histo_configs/classification/BaseMIL.py [hunk 5]
```diff
@@ -44,8 +44,12 @@ class BaseMIL(LightningContainer):
...
+    enconding_chunk_size: int = param.Integer(0, doc="If > 0 performs encoding in chunks, by loading"
```
Comment (ant0nsc) [inline InnerEye/ML/configs/histo_configs/classification/BaseMIL.py:51]:
```suggestion
    encoding_chunk_size: int = param.Integer(0, doc="If > 0 performs encoding in chunks, by loading"
```

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py [hunk 6]
```diff
@@ -39,6 +39,9 @@ def __init__(self, **kwargs: Any) -> None:
         default_kwargs = dict(
             # declared in BaseMIL:
             pooling_type=GatedAttentionLayer.__name__,
+            # average number of tiles is 56 for PANDA
+            enconding_chunk_size=60,
```
Comment (ant0nsc) [inline InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py:43]:
```suggestion
            encoding_chunk_size=60,
```

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py [hunk 7]
```diff
@@ -101,7 +106,7 @@ def get_data_module(self) -> PandaTilesDataModule:
         transform = Compose(
             [
                 LoadTilesBatchd(image_key, progress=True),
-                EncodeTilesBatchd(image_key, self.encoder),
+                EncodeTilesBatchd(image_key, self.encoder, chunk_size=self.enconding_chunk_size),
```
Comment (ant0nsc) [inline InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py:109]:
```suggestion
                EncodeTilesBatchd(image_key, self.encoder, chunk_size=self.encoding_chunk_size),
```

File: InnerEye/ML/deep_learning_config.py [hunk 8]
```diff
@@ -475,6 +475,7 @@ def logs_folder(self) -> Path:
     @property
     def checkpoint_folder(self) -> Path:
         """Gets the full path in which the model checkpoints should be stored during training."""
+        print(f"Expected Checkpoint path {self.outputs_folder / CHECKPOINT_FOLDER}")
```
Comment (ant0nsc) [inline InnerEye/ML/deep_learning_config.py:478]: This print statement is very noisy, and clutters the output. We are using this property a LOT. Remove, or at least reduce it to rank zero only?
Also, I generally find it not ideal if properties have side effects.

**Review by ant0nsc — COMMENTED** [review]
Looking good, just a few minor comments

File: InnerEye/ML/deep_learning_config.py (hunk 8 above)
Comment (vale-salvatelli) [reply InnerEye/ML/deep_learning_config.py:478]: Removed. This was a leftover from debugging, thanks for spotting

File: InnerEye/ML/Histopathology/datamodules/base_module.py (hunk 3 above)
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py:111]: changed

File: InnerEye/ML/Histopathology/datamodules/base_module.py (hunk 2 above)
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py:103]: it is the standard abbreviation for pytorch when saving in pickle format - I can revert to pickle if more clear?

File: InnerEye/ML/configs/histo_configs/classification/BaseMIL.py (hunk 4 above)
Comment (vale-salvatelli) [reply InnerEye/ML/configs/histo_configs/classification/BaseMIL.py:50]: I was trying to maintain the same pattern we already had for CacheMode but I understand is not beautiful

File: InnerEye/ML/Histopathology/datamodules/base_module.py [hunk 9]
```diff
@@ -46,16 +50,25 @@ def __init__(self, root_path: Path, max_bag_size: int = 0, batch_size: int = 1,
...
         transforms up to the first randomised one should be computed only once and reused in
         subsequent iterations:
-          - `MEMORY`: the entire transformed dataset is kept in memory for fastest access;
-          - `DISK`: each transformed sample is saved to disk and loaded on-demand;
-          - `NONE` (default): no caching is performed.
-        :param save_precache: Whether to pre-cache the entire transformed dataset upfront and save
+          - `MEMORY`: MONAI CacheDataset is used, the entire transformed dataset is kept in memory for fastest access;
+          - `DISK`: MONAI PersistentDataset is used, each transformed sample is saved to disk and loaded on-demand;
+          - `NONE` (default): standard MONAI dataset is used, no caching is performed.
+        :param precache_location: Whether to pre-cache the entire transformed dataset upfront and save
```
Comment (dccastro) [inline InnerEye/ML/Histopathology/datamodules/base_module.py:56]: Let's clarify that this applies only to memory caching (cf. comment on disk caching below). Also, I'd maybe rephrase the description less as a Boolean flag ("Whether [...]").

File: InnerEye/ML/Histopathology/datamodules/base_module.py [hunk 10]
```diff
@@ -46,16 +50,25 @@ def __init__(self, root_path: Path, max_bag_size: int = 0, batch_size: int = 1,
...
         it to disk. This is done once in `prepare_data()` only on the local rank-0 process, so
-        multiple processes can afterwards access the same cache without contention in DDP settings.
+        multiple processes can afterwards access the same cache without contention in DDP settings. 3 options are
+        available:
+          - `CPU`: each transformed sample is saved to disk and loaded into CPU memory on-demand;
+          - `GPU`: each transformed sample is saved to disk and loaded into GPU memory on-demand;
```
Comment (dccastro) [inline InnerEye/ML/Histopathology/datamodules/base_module.py:61]: Given that this argument applies to `MEMORY` mode only, consider rephrasing these descriptions to clarify that the whole dataset is saved and loaded at once, rather than on-demand.

File: CHANGELOG.md [hunk 11]
```diff
@@ -18,6 +18,7 @@ created.
   GPU utilization via Lightning's `GpuStatsMonitor`, switch `monitor_loading` to check batch loading times via
   `BatchTimeCallback`, and `pl_profiler` to turn on the Lightning profiler (`simple`, `advanced`, or `pytorch`)
 - ([#544](https://github.com/microsoft/InnerEye-DeepLearning/pull/544)) Add documentation for segmentation model evaluation.
+- ([#637](https://github.com/microsoft/InnerEye-DeepLearning/pull/637)) Add option to encode in chunks and to load pre-cached dataset to the histo pipeline.
```
Comment (dccastro) [inline CHANGELOG.md:21]: Minor suggestion to clarify that the "option [...] to load pre-cached dataset" refers to CPU vs GPU memory.

File: InnerEye/ML/Histopathology/datamodules/base_module.py [hunk 12]
```diff
@@ -46,16 +50,25 @@ def __init__(self, root_path: Path, max_bag_size: int = 0, batch_size: int = 1,
...
+          - `NONE (default)`: no pre-cache is performed;
         :param cache_dir: The directory onto which to cache data if caching is enabled.
         :param number_of_cross_validation_splits: Number of folds to perform.
         :param cross_validation_split_index: Index of the cross validation split to be performed.
         """
+        if precache_location == CacheLocation.NONE:
+            save_precache = None
+        else:
+            save_precache = precache_location
```
[the remark this answers is not in the document]
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py:70]: I was trying to keep low the number of parameters relative to cache to avoid further confusion between the different options. the only reason why here I redefined a save_precache was to make the assert statement a bit more compact. Let me remove, I see this confused both you and Anton

File: InnerEye/ML/Histopathology/datamodules/base_module.py [hunk 13]
```diff
@@ -87,17 +100,20 @@ def prepare_data(self) -> None:
...
     def _load_dataset(self, tiles_dataset: TilesDataset, stage: str, shuffle: bool) -> Dataset:
-        dataset_pickle_path = self._dataset_pickle_path(stage)
+        dataset_pt_path = self._dataset_pt_path(stage)
 
-        if dataset_pickle_path and dataset_pickle_path.exists():
-            with dataset_pickle_path.open('rb') as f:
-                return pickle.load(f)
+        if dataset_pt_path and dataset_pt_path.exists():
+            # torch.load will reload on GPU by default, same device it was saved from
+            memory_location = torch.device('cpu') if self.save_precache == CacheLocation.CPU else None
```
[the remark this answers is not in the document]
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py:113]: good observation - I'll change things to make this clear

[the remark this answers is not in the document]
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py:113]: two reasons:
1 - no need to, torch.load will reload on GPU by default (see comment above)
2 - we would need to specify the device, I understand the correct syntax is `cuda:device`, that means I would need to propagate the device info here

[the remark this answers is not in the document]
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py:113]: Changed. To be noted: this parameter still regulates in the `CacheMode.DISK` case if we want to precache or not. Tried to make the docs as explicit as possible

File: InnerEye/ML/Histopathology/datamodules/base_module.py (hunk 9 above)
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py:56]: done

File: InnerEye/ML/Histopathology/datamodules/base_module.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py:70]: save_precache removed

File: InnerEye/ML/configs/histo_configs/classification/BaseMIL.py (hunk 4 above)
Comment (vale-salvatelli) [reply InnerEye/ML/configs/histo_configs/classification/BaseMIL.py:50]: I left this untouched but removed the horrible construct you commented on above

File: InnerEye/ML/Histopathology/datamodules/base_module.py (hunk 1 above)
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py:71]: this is now removed

File: InnerEye/ML/Histopathology/datamodules/base_module.py (hunk 2 above)
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py:103]: revert to pickle for more clarity

File: InnerEye/ML/Histopathology/datamodules/base_module.py (hunk 10 above)
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py:61]: done

File: InnerEye/ML/Histopathology/datamodules/base_module.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dccastro) [reply InnerEye/ML/Histopathology/datamodules/base_module.py]: 1. It will reload on whatever original device. Note that this class is generic w.r.t. the `transform`, meaning the data to cache might live in CPU. In this case, the user who specified `CacheLocation.GPU` would be surprised that their data is not loaded on GPU.
2. Just `'cuda'` without an index maps to the default GPU: https://pytorch.org/docs/stable/notes/cuda.html

[the remark this answers is not in the document]
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py]: 1. I hear you, I haven't planned for the option where the result of the transformation is not already on gpu. Let me capture that case too.
2. I know we can use `cuda` for a generic device but here the saving will happen on one single device and when we reload we want each process to reload in a different device, if we just specify `cuda` I understand it will instead reload always to the same, whatever device was used before saving. [https://pytorch.org/docs/stable/generated/torch.load.html](https://pytorch.org/docs/stable/generated/torch.load.html).  Am I wrong?

[the remark this answers is not in the document]
Comment (dccastro) [reply InnerEye/ML/Histopathology/datamodules/base_module.py]: 1. Sounds good.
2. You're right. Maybe a more flexible and robust option is to let the user directly specify the device (`'cpu'`, `'cuda:xxx'`, or `None`)?

[the remark this answers is not in the document]
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/datamodules/base_module.py]: Discussed at voice. In the end I chose not to support the option save in cpu - load to gpu. The new variable name should convey clearly to the user what are the possibilities.

File: Tests/ML/histopathology/models/test_transforms.py [hunk 14]
```diff
@@ -126,8 +126,10 @@ def test_cached_loading(tmp_path: Path) -> None:
 
 @pytest.mark.skipif(not os.path.isdir(TCGA_CRCK_DATASET_DIR),
                     reason="TCGA-CRCk tiles dataset is unavailable")
-@pytest.mark.parametrize('use_gpu', [False, True])
-def test_encode_tiles(tmp_path: Path, use_gpu: bool) -> None:
+@pytest.mark.parametrize('use_gpu , chunk_size',
+                         [(False, 0), (False, 2), (True, 0), (True, 2)]
+                         )
+def test_encode_tiles(tmp_path: Path, use_gpu: bool, chunk_size: int) -> None:
```
Comment (ant0nsc) [inline Tests/ML/histopathology/models/test_transforms.py:132]: This test relies on datasets being mounted, and this is only the case on some of people's individual VMs. This makes the test not particularly useful. We should talk about a better way of creating datasets for testing - mocking being one obvious option.

**Review by ant0nsc — APPROVED** [review]

**Review by dccastro — APPROVED** [review]

File: Tests/ML/histopathology/models/test_transforms.py (hunk 14 above)
Comment (vale-salvatelli) [reply Tests/ML/histopathology/models/test_transforms.py:132]: Agreed with you - most of the tests for histo relies on this strategy
